### DISCUSSION TITLE
Have DataTables display widgets in cells

### DISCUSSION THREAD

**Discussion by SebastienSyd** [Textualize/textual#2152] [Ideas]
I am using DataTables and I cannot display widgets in cells:
![image](https://user-images.githubusercontent.com/28516084/228206122-16320824-3caf-4b9a-952b-2c49fe90cbb5.png)
In the above example I tried to display a link with `Static("[@click='app.bell']link[/]")` (just for testing)

The idea would be to display any Static content (Links, Labels, etc.) or even buttons so that we can trigger an action based on the row the button is located

**Comment by davep** [comment]
The `DataTable` is designed around displaying Rich renderables, which helps make it as efficient as possible. We don't have any plans any time soon for having it contain widgets.

If you want to links and the like in the DataTable you should be able to do that without using a `Static` or a `Label` as the cells in a `DataTable` will contain a renderable anyway.

**Reply by SebastienSyd** [reply]
thanks @davep for your answer. I will give it a try with the links

**Reply by SebastienSyd** [reply]
Hi @davep I tried and that didn't work
I tried with `[link=https://www.willmcgugan.com]blog[/link]` for example, but it renders exactly the text without a link

Do you have a concrete example? 🙏

**Reply by SebastienSyd** [reply]
![image](https://user-images.githubusercontent.com/28516084/228914314-4664eaff-149e-4785-822b-44b2ad999406.png)

**Reply by davep** [reply]
I was testing something else earlier today this this:

```py
from textual.app        import App, ComposeResult
from textual.widgets    import Header, Footer, DataTable

class DTLinksApp( App[ None ] ):

    CSS = """
    DataTable {
        height: 1fr;
    }
    """

    def compose( self ) -> ComposeResult:
        yield Header()
        yield DataTable()
        yield Footer()

    def on_mount( self ) -> None:
        dt = self.query_one( DataTable )
        dt.focus()
        dt.add_columns( "Title", "URL Markup", "No Markup" )
        dt.add_rows( [ (
            f"Relevant XKCD {n}",
            f"[link]https://xkcd.com/{n}/[/]",
            f"https://xkcd.com/{n}/"
        ) for n in range( 500 ) ] )

if __name__ == "__main__":
    DTLinksApp().run()
```

and the links were clickable in every terminal I tried.

![Screenshot 2023-03-30 at 18 23 33](https://user-images.githubusercontent.com/28237/228915950-9f83a1de-9bef-42b0-9b5e-3de7b4fe6a8d.png)

and if I use a link with a label, that works too:

```python
from textual.app        import App, ComposeResult
from textual.widgets    import Header, Footer, DataTable

class DTLinksApp( App[ None ] ):

    CSS = """
    DataTable {
        height: 1fr;
    }
    """

    def compose( self ) -> ComposeResult:
        yield Header()
        yield DataTable()
        yield Footer()

    def on_mount( self ) -> None:
        dt = self.query_one( DataTable )
        dt.focus()
        dt.add_columns( "Title", "URL Markup", "No Markup" )
        dt.add_rows( [ (
            f"Relevant XKCD {n}",
            f"[link=https://xkcd.com/{n}/]XKCD {n}[/]",
            f"https://xkcd.com/{n}/"
        ) for n in range( 500 ) ] )

if __name__ == "__main__":
    DTLinksApp().run()
```

![Screenshot 2023-03-30 at 18 26 33](https://user-images.githubusercontent.com/28237/228916560-bc9e7f71-23b1-45f5-b762-a2c0ebc237e9.png)

What version of Textual are you using?

**Reply by SebastienSyd** [reply]
I am using `textual==0.16.0`

**Reply by davep** [reply]
0.17 enabled processing of markup by default, so that will be the difference I believe. Either upgrade or if you need to remain on 0.16 for now use Text.from_markup from Rich to turn the links into renderables.

**Reply by SebastienSyd** [reply]
Thank you! Will upgrade tomorrow and test this out.
I greatly appreciate your time!

**Reply by davep** [reply]
No worries, glad to help. Let me know how you get on. 🙂

**Reply by SebastienSyd** [reply]
works like a charm with the latest version 👍

**Comment by IPdotSetAF** [comment]
Is it possible to extend the datagrid widget to do support widgets?
i need to display a small graph
